packages/api/src/folder/controller.test.ts

```diff
@@ -0,0 +1,177 @@
+
+        next();
+      }
+    );
+    await clearDatabase();
+    await setupDatabase();
+  });
+
+  afterEach(async () => {
+    await clearDatabase();
+    jest.restoreAllMocks();
+    jest.clearAllMocks();
+  });
+
+  test("expect an empty query to cause a 400 error", async () => {
+    await agent.patch("/api/v2/folder/1").send({}).expect(400);
+  });
+
+  test("expect non existent folder to cause a 404 error", async () => {
+    await agent
+      .patch("/api/v2/folder/111111111")
+      .send({ displayDate: "2024-09-26T15:09:52.000Z" })
+      .expect(404);
+  });
+
+  test("expect request to have an email from auth token if an auth token exists", async () => {
+    (extractUserEmailFromAuthToken as jest.Mock).mockImplementation(
+      (req: Request, __, next: NextFunction) => {
+        (req.body as { emailFromAuthToken: string }).emailFromAuthToken =
+          "not an email";
+        next();
+      }
+    );
+
+    await agent.patch("/api/v2/folder/1").expect(400);
+  });
+
+  test("expect displayDate is updated", async () => {
+    await agent
+      .patch("/api/v2/folder/1")
+      .send({ displayDate: "2024-09-26T15:09:52" })
+      .expect(200);
+
+    const result = await db.query(
+      "SELECT to_char(displaydt, 'YYYY-MM-DD HH24:MI:SS') as displaydt FROM folder WHERE folderid = :folderId",
+      {
+        folderId: 1,
+      }
+    );
+
+    expect(result.rows[0]).toEqual({ displaydt: "2024-09-26 15:09:52" });
+  });
+
+  test("expect displayDate is updated when set to null", async () => {
+    await agent
+      .patch("/api/v2/folder/1")
+      .send({ displayDate: null })
+      .expect(200);
+
+    const result = await db.query(
+      "SELECT displaydt FROM folder WHERE folderid = :folderId",
+      {
+        folderId: 1,
+      }
+    );
+
+    expect(result.rows[0]).toStrictEqual({ displaydt: null });
+  });
+
+  test("expect 400 error if display date is wrong type", async () => {
+    await agent
+      .patch("/api/v2/folder/1")
+      .send({
+        displayDate: false
+      })
+      .expect(400);
+  });
+
+  test("expect to log error and return 500 if database update fails", async () => {
+    const testError = new Error("test error");
+    const spy = jest.spyOn(db, "query").mockImplementation(async () => {
+      throw testError;
+    });
+
+    await agent
+      .patch("/api/v2/folder/1")
+      .send({ displayDate: "2024-09-26T15:09:52.000Z" })
+      .expect(500);
+    spy.mockRestore();
+
+    expect(logger.error).toHaveBeenCalledWith(testError);
+  });
+
+  test("expect to log error and return 500 if database update is ok but the database select fails", async () => {
+    const testError = new Error("test error");
+    const originalBehavior = db.sql.bind(db);
+    jest.spyOn(db, "sql").mockImplementation((name, ...params) => {
+      if (name === 'folder.queries.get_folder_by_id') {
+        throw testError;
+      }
+
+      return originalBehavior(name, params)
+    });
```

Comment:
`mockImplementationOnce()` is what I usually use for these cases where I need the first DB call to succeed an a later one to fail: https://github.com/PermanentOrg/stela/blob/8130f222a325214ffef2f6af1914a138f059c46a/packages/api/src/billing/controller.test.ts#L615

Reply:
I did that, but I'm seeing some issues with this approach.

1. I would like to mock just one sql call, I want the .sql call to actually reach the database because I want to check if the update actually happened in the DB. I did a .sql call inside the test to check that, but it will just return my mocked response.

2. This is highly dependent on the query order sequence. Adding/Removing one sql call at some point in our code may break a lot of tests and we would have to redo the mocks based on the new order of the sql queries.

Isn't there a way to mock the response based on the function params?

Reply:
Though I don't know if that will provide the behavior you're looking for of "return a mocked response for this set of arguments, otherwise behave as normal". I think it would be reasonable, within this test, to check that the correct update call to db.sql is made, rather than checking that the database itself is updated. Other tests ensure that the update call correctly issued will result in an update.

Reply:
Pushing the changes without jest-when, but looking into how can we use it. I think it's the more sustainable option.